mantidproject/mantid
# Fix default Q binning in DirectILLReduction

The default Q binning in `DirectILLReduction` was larger than what was convenient for the users. This PR fixes the problem by making the 'rounding' of the bin width a bit smarter.

This PR also fixes a 'leaking' temporary workspace when the optional *OutputSofThetaEnergyWorkspace* is set.

**To test:**

Make sure you have unit test data in Mantid's data search directories. Then run the following script:
```python
DirectILLCollectData('ILL/IN4/084447.nxs', OutputWorkspace='ws')
DirectILLReduction('ws', OutputWorkspace='SofQW')
```

The logs should say `Binned momentum transfer axis to bin width 0.035 A-1.`, and the X binning should correspond.

Fixes #22147.

-----
As part of this work the branch for #21741 was merged in, since it detected NaN values being used during the unit tests. 

If a user manages to create x bins with the value of `NaN` or `infinity` the rebin algorithm will enter an infinite loop in `VectorHelper::createAxisFromRebinParams`.

This is because the current x value turns into an invalid number, which fails all comparisons. Therefore the while loop will not terminate.

**To test:**
Run the following script:
```
CreateSampleWorkspace(OutputWorkspace='ws', WorkspaceType='Histogram', BankPixelWidth=3, PixelSpacing=0.1, NumBanks=1, Random=True, XMin=0, XMax=20, BinWidth=2, XUnit="Wavelength")
MoveInstrumentComponent(Workspace='ws', ComponentName='bank1', Y=0.026181)
ConvertUnits(InputWorkspace='ws',OutputWorkspace='ws_Q', Target='MomentumTransfer')
RebinToWorkspace(WorkspaceToRebin='ws_Q',WorkspaceToMatch='ws_Q', OutputWorkspace='ws_Q_rebinned')
```

It should not run infinitely and instead emit an error message and stop.

Fixes #21741 

-----

**Release Notes** 

Direct inelastic release notes were updated accordingly.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
